src/test/java/org/folio/rest/api/RequestBatchAPITest.java

```diff
@@ -0,0 +1,192 @@
+
+    // 2nd: Specify same positions for the requests, so constraint violation happens
+    firstRequest.put("position", 1);
+    secondRequest.put("position", 1);
+
+    CompletableFuture<JsonResponse> putCompleted = new CompletableFuture<>();
+    client.put(batchRequestStorageUrl(),
+      buildBatchUpdateRequest(firstRequest, secondRequest),
+      TENANT_ID,
+      json(putCompleted)
+    );
+
+    JsonResponse response = putCompleted.get(5, TimeUnit.SECONDS);
+    assertThat(response.getStatusCode(), is(422));
+
+    JsonArray errors = response.getJson().getJsonArray("errors");
+    assertThat(errors.size(), is(1));
+    assertThat(errors.getJsonObject(0).getString("message"),
+      is("Cannot have more than one request with the same position in the queue"));
+
+    // 3rd: Retrieve requests from DB and assert that changes have not been applied.
+    JsonObject requestsFromDb = getAllRequestsForItem(itemId);
+
+    JsonObject firstRequestFromDb = requestsFromDb.getJsonArray("requests").getJsonObject(0);
+    JsonObject secondRequestFromDb = requestsFromDb.getJsonArray("requests").getJsonObject(1);
+
+    assertThat(requestsFromDb.getInteger("totalRecords"), is(2));
+    assertThat(firstRequestFromDb, is(firstRequest));
+    // 2nd RQ must have it's initial position - 2nd
+    assertThat(secondRequestFromDb, is(secondRequest.put("position", 2)));
+  }
+
+  @Test
+  public void canRollbackTransactionServerError() throws Exception {
+    UUID itemId = UUID.randomUUID();
+
+    // 1st create some sample requests for an item
+    JsonObject firstRequest = createRequestForItemAtPosition(itemId, 1);
+    JsonObject secondRequest = createRequestForItemAtPosition(itemId, 2);
+
+    // 2nd: Remove ID for the request, which causes an exception and transaction to be reverted
+    JsonObject firstRequestCopy = firstRequest.copy();
+
+    firstRequestCopy.remove("id");
+    secondRequest.put("position", 10);
+
+    CompletableFuture<Response> putCompleted = new CompletableFuture<>();
+    client.put(batchRequestStorageUrl(),
+      buildBatchUpdateRequest(firstRequestCopy, secondRequest),
+      TENANT_ID,
+      empty(putCompleted)
+    );
+
+    // Expect server error due to Exception
+    Response response = putCompleted.get(5, TimeUnit.SECONDS);
+    assertThat(response.getStatusCode(), is(500));
+
+    // 3rd: Retrieve requests from DB and assert that changes have not been applied.
+    JsonObject requestsFromDb = getAllRequestsForItem(itemId);
+
+    JsonObject firstRequestFromDb = requestsFromDb.getJsonArray("requests").getJsonObject(0);
+    JsonObject secondRequestFromDb = requestsFromDb.getJsonArray("requests").getJsonObject(1);
+
+    assertThat(requestsFromDb.getInteger("totalRecords"), is(2));
+    assertThat(firstRequestFromDb, is(firstRequest));
+    // 2nd RQ must have it's initial position - 2nd
+    assertThat(secondRequestFromDb, is(secondRequest.put("position", 2)));
```

Comment:
I think it is clearer to get the position from the JSON and assert against the raw property, rather than assert against a mutated copy of entire JSON. 

Why is the entire JSON is being asserted on?

Reply:
The main purpose of asserting entire JSON to make sure that not even position has not been changed but all other fields are not modified.

Reply:
@bohdan-suprun thanks for investigating the deferred constraints

Reply:
I can understand that. Maybe there is a way to make that clearer, maybe by not doing the position mutation inline, or using the original representation prior to the attempted operation?

Reply:
Outdated.